src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+            for pline in self.cleanuplines:
+                try:
+                    pline.run(wflow)
+                except Exception as ex:
+                    msg = "%s: General error from pipeline. Error: \n%s. "
+                    msg += "\nWill retry again in the next cycle."
+                    self.logger.exception(msg, pline.name, str(ex))
+                    break
+                if wflow['CleanupStatus'][pline.name]:
+                    self.wfCounters['cleaned'][pline.name] += 1
+
+        # Archive:
+        try:
+            self.plineArchive.run(wflow)
+            self.wfCounters['archived'] += 1
+        except MSRuleCleanerArchival as ex:
+            msg = "%s: Archival Error: %s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.error(msg, wflow['PlineMarkers'][-1], ex.myMessage)
+        except Exception as ex:
+            msg = "%s General error from pipeline. Error: \n%s. "
+            msg += "\nWill retry again in the next cycle."
+            self.logger.exception(msg, wflow['PlineMarkers'][-1], str(ex))
+        return wflow
+
+    def setPlineMarker(self, wflow, pName=None):
+        """
+        A function intended to mark which is the pipeline currently working
+        on the workflow. It is supposed to be called always as a first function
+        in the pipeline.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :param  pName:   The name of the functional pipeline
+        :return wflow:
+        """
+        # NOTE: The current functional pipeline MUST always be appended at the
+        #       end of the 'PlineMarkers' list
+
+        # First check if we can safely get rid of the default:
+        if len(wflow['PlineMarkers']) == 1 \
+           and wflow['PlineMarkers'][-1] is None \
+           and pName is not None:
+            wflow['PlineMarkers'].pop()
+
+        # Then push our current value into the markers list:
+        wflow['PlineMarkers'].append(pName)
+        return wflow
+
+    def _checkClean(self, wflow):
+        """
+        An auxiliary function used to only check the temporary cleanup status.
+        It basically takes the pipelines registered in 'PlineMarkers' that have
+        already worked on the workflow as a mask and applies this mask over
+        the set of flags in the 'CleanupStatus' field and then reduces the
+        result to a single bool value
+        """
+        # NOTE: This is one of the few functions taking a workflow as an argument
+        #       but returning a bool, since it is an auxiliary function and is not
+        #       supposed to be called as a standalone function in a pipeline.
+        # NOTE: `all([]) == True`, ergo all the 'rejected' && 'aborted-completed' workflows
+        #       are also counted as properly cleaned and can trigger archival later
+        cleanStatus = all([wflow['CleanupStatus'][key]
```

Comment:
In addition to, this would be more readable if written in multiple lines.
I also think this is somehow a fragile construction, where we need to ensure that all the necessary pipelines get reflected in the `CleanupStatus` structure.

What came to my mind was that, all the `CleanupStatus` key/value pairs should be updated according to the status (read pipelines that need to be executed) and before a workflow gets archived, we need to make sure all of those are `True`.

Reply:
This  was the initial construction and was working wrong. It was giving false results, because that way one reads all the flags but does not take into account that based on the conditions from the upper logic (like workflow status, type of rules to be cleaned, parentage resolved flag and so on) a different set of pipelines are working on the workflow. So not only the counters were reported wrong but also archival was triggered for uncleaned workflows.  
This comment of yours is 1 to 1 related to the previous one, so leaving this one open too, because of the ongoing work.

Reply:
Perhaps something else to consider here is, to make sure that if any pipeline hits a bad exception, that we still have that pipeline writing its name and status to this workflow structure. Otherwise we could have things going to archive because the archival pipeline isn't aware of a broken pipeline in the chain.
I fail to see if this is a legit concern or if the code has a protection for that though.

Reply:
It is a legit concern, Yes. And that's  why I keep all default values to `False`. I usually try to watch out for such behavior. So what can ever happen are the following two scenarios:
   * The pipeline is marked (by having the very first function called to be `stPlineMarkers()`), but fails somewhere down the road... then what ever exception is thrown will be caught as a 'General Error' by the top level caller - `_dispatcher()` and the archival call will still be triggered, since it is the next step in the `_dispatcher()` but the flags would have already be set to `False` and no archival will happen.
   * The pipeline fails completely and it cannot even set it's markers then the whole `_dispatcher()` blows up at line 323 (currently) when it tries to increase an un-existing counter with: 
```                   self.wfCounters['cleaned'][pline.name] += 1 ```

I have not found a third sequence of events so far.